### PR TITLE
Fixes #27891 - Add description field to templates

### PR DESCRIPTION
**Pull request by ofedoren**
Adds description field to templates.
ScreenShot-1568909963756

Chages are in UI and API (probably will require a PR to hammer to add the field there as well).

### CODE REVIEW

**Review by ezr-ondrej — DISMISSED** [review]
Code-wise LGTM.

**Comment by ezr-ondrej** [comment]
The code seems good, just we see it only while editing in UI, but that can be fixed later.
But the use-case suggested on the screenshot is not what I would expect from such field. I would like to see such not on generating (generally using the template) I guess.
I don't like editable field being a warning against editing another field.

**Comment by ares** [comment]
So we may have two use-cases here. General purpose description field and special meening warnings, examples from community templates:

description field (aka usage):
* https://github.com/theforeman/community-templates/blob/develop/provisioning_templates/provision/kickstart_default.erb#L11-L31
* https://github.com/theforeman/community-templates/blob/develop/provisioning_templates/snippet/redhat_register.erb#L7-L67
* https://github.com/theforeman/community-templates/blob/develop/provisioning_templates/provision/kickstart_ovirt.erb#L9-L37

warning:
* the one listed here - ansible inventory
* https://github.com/theforeman/community-templates/blob/develop/provisioning_templates/PXELinux/pxelinux_global_default.erb#L6

Do we need 2 different fields? Should warning be editable or should it be set only when importing the template according to metadata?

**Comment by lzap** [comment]
I don't think we need two of those. Users who edit it have been warned...

**Comment by ofedoren** [comment]
> I don't think we need two of those. Users who edit it have been warned...

You are pretty optimistic with this approach :D

I agree if we could split it to two fields. 
 - The basic description could be changed and is shown while editing and generating.
 - The warning message could be set once (via import or at creating without possibility to change) and then is shown only when editing the template.

@ares, @ezr-ondrej, @lzap, what do you think?

**Comment by ezr-ondrej** [comment]
I would still rather see the locking introduced in report templates. We already have that in provision templates. Our users knows it, so why to introduce warning?

**Comment by lzap** [comment]
> You are pretty optimistic with this approach :D

Maybe I am just explaining my opinion badly. Let me rephrase.

If an user edit a single textbox, he/she had to see the original text (some kind of warning). Therefore it was conscious decision to erase it. Problem solved.

Everything has to have its own price, I would like to see more reasons to actually create *another* column in a database. Let's try to keep things simple.

**Comment by ezr-ondrej** [comment]
> If an user edit a single textbox, he/she had to see the original text (some kind of warning). Therefore it was conscious decision to erase it. Problem solved.

I don't agree it's necessarily conscious choice. And even if it is, it is choice of one user in multi user system :thinking: 
IMO description is too general field to be used as a warning. Such design is screaming "designed by an Software engineer" to me :man_shrugging:

**Comment by ares** [comment]
> I would still rather see the locking introduced in report templates. We already have that in provision templates. Our users knows it, so why to introduce warning?

I'm not sure I follow, we already have locking in report templates and default report templates are locked right?

I think it would be good to have two separate fields. One is for generic description or comment that people can use for their notes. E.g. "this is a PoC, feel free to modify" or "Don't change without asking Marek", "This is used to provision ACME DB server".

Then some templates have a special meaning, such as "This template is used for rendering global PXE menu, do not rename it" or "This report template is used to generate ansible inventory, do not rename it".

This does not have to be something, user can edit, but would be good for user to see as a warning in the app. We don't have to store it in DB as we'd have to translate that string anyway. But from user perspective, I don't want to read about special meening of the template somewhere in the manual. We can separate it from this PR and maybe implement something better, such as choosing the specific template in setting and display warning based on that setting. The generic description/comment is good addition regardless.

**Comment by ares** [comment]
So :+1: for this, I haven't tested myself, but we should make sure it gets populated when the template is imported through API or foreman_templates plugin. If that's the case, we can merge.

**Comment by ekohl** [comment]
> Then some templates have a special meaning, such as "This template is used for rendering global PXE menu, do not rename it" or "This report template is used to generate ansible inventory, do not rename it".

Don't we have a mechanism to select the template to use? IIRC there's some UI for that. If that works, it could autodetect this and show a generic notification about that.

**Comment by ares** [comment]
> Don't we have a mechanism to select the template to use? IIRC there's some UI for that. If that works, it could autodetect this and show a generic notification about that.

not all of templates with a special meaning IIRC, but yeah, some can be configured through settings today

**Comment by ofedoren** [comment]
I've updated the PR, so partition tables can be created from API with description as well. UI works for job templates, but API params/views can't be updated within the core, can they?

> So  for this, I haven't tested myself, but we should make sure it gets populated when the template is imported through API or foreman_templates plugin. If that's the case, we can merge.

@ares, exporting/importing seems to be working now as well.

**Review by ares — CHANGES_REQUESTED** [review]
@ofedoren you're right about job templates. This will need a follow up PR to REX to add description paramt to REX API, can you please open that?

I've tested the PR and it looks and works well. There are 2 things that should be still updated:
1) if the template is locked, other fields in the form are locked, description remains unlocked (validation prevents from saving but the field should be disabled)
2) the description will be often just blank white box while the editor is the most important part of the form, could we move description field under the editor?

**Comment by ofedoren** [comment]
@ares, done.

UPD: PR to REX plugin: https://github.com/theforeman/foreman_remote_execution/pull/436

**Review by ares — APPROVED** [review]
APJ (ack pending jenkins)

**Comment by ares** [comment]
I suppose you already know, but we'll also need to update hammer-cli-foreman and REX plugin to fully take advantage of this right? Anyway, works nice for me, test failure unrelated, merging!

**Comment by ares** [comment]
Thanks everyone for the discussion above, I went ahead with getting the description field in as that is generally useful and does not prevent us to add warning texts based on templates selected via settings. Until we get there, we can use descriptions for this purpose too. It would be awesome if someone could go over community templates and modify ERB comments to descriptions.
